sympy/solvers/solveset.py

```diff
@@ -1117,3 +1120,525 @@ def linsolve(system, *symbols):
+    {(-1, 1)}
+    >>> substitution([x +y], [x], [{y: 1}], [y], set([x + 1]), [y, x])
+    EmptySet()
+    >>> substitution([x +y], [x], [{y: 1}], [y], set([x - 1]), [y, x])
+    {(1, -1)}
+
+    """
+
+    from sympy.core.compatibility import ordered, default_sort_key
+    from sympy import Complement
+    from sympy.core.containers import Dict
+
+    def _unsolved_syms(eq, sort=False):
+        # get the unsolved symbol.
+        free = eq.free_symbols
+        unsolved = (free - set(known_symbols)) & set(all_symbols)
+        if sort:
+            unsolved = list(unsolved)
+            unsolved.sort(key=default_sort_key)
+        return unsolved
+
+    # sort such that equation with the fewest potential symbols is first.
+    # means eq with less variable first.
+    complements = {}
+    for eq in ordered(system, lambda _: len(_unsolved_syms(_))):
+        newresult = []
+        got_s = set()  # symbols solved in one iteration
+        hit = False
+        for r in result:
+            # update eq with everything that is known so far
+            eq2 = eq.subs(r)
+            unsolved_syms = _unsolved_syms(eq2, sort=True)
+            if not unsolved_syms:
+                if r:
+                    newresult.append(r)
+                break  # skip as it's independent of desired symbols
+            for s in unsolved_syms:
+                not_solvable = False
+                soln_imageset = None
+                try:
+                    soln = solveset_real(eq2, s)
+                    # If we want to add the complex solution also then
+                    # do the same with soln_complex. But it sometimes
+                    # we get complicated solution, so not adding it always
+                    # soln_complex = solveset_complex(eq2, s)
+                    if not soln:
+                        # If there is no real solution then check for
+                        # complex solution like old solve.
+                        soln = solveset_complex(eq2, s)
+                except NotImplementedError:
+                    continue
+                if isinstance(soln, Complement):
+                    # extract solution and complement
+                    complements[s] = soln.args[1]
+                    soln = soln.args[0]
+                    # complement should be added at the end
+                if isinstance(soln, ImageSet):
+                    soln_imageset = soln
+                    soln = FiniteSet(soln.lamda.expr)
+                elif isinstance(soln, ConditionSet):
+                    soln = FiniteSet()
+                    # dont do `continue` we may get soln
+                    # in terms of other symbol(s)
+                    not_solvable = True
+                elif isinstance(soln, Intersection):
+                    # sometimes solveset returns Intersection with S.Real
+                    # when #11174 is fixed,then this line should be removed
+                    soln = soln.args[1]
```

Comment:
Because of this issue https://github.com/sympy/sympy/issues/11174 I added these lines. 
eg.
```
In [2]: solveset(x + y - 2, x ,S.Reals)
Out[2]: (-∞, ∞) ∩ {y}

In [3]: x, y, z = symbols('x, y, z', real = True)

In [4]: solveset(x + y - 2, x ,S.Reals)
Out[4]: {-y + 2}
```
So I didn't edited this because of the issue. 
My actual intention was if soln is like `Intersection( FiniteSet(-y), Interval(-oo, 0))`

```
In [25]:  Intersection(Interval(-oo, 0), FiniteSet(-y))
Out[25]: [0, ∞) ∩ {y}

In [26]:  Intersection( FiniteSet(-y), Interval(-oo, 0))
Out[26]: [0, ∞) ∩ {y}

In [27]:  Intersection( FiniteSet(-y), Interval(-oo, 0)).args[0]
Out[27]: [0, ∞)

In [28]:  Intersection( FiniteSet(-y), Interval(-oo, 0)).args[1]
Out[28]: {y} # choose this for further steps

```
and in final soln add with intersection.

I will change it.